### PR TITLE
Harden against failure to resolve packages or produce assets cache

### LINKED ISSUE #2044: The "ResolvePackageAssets" task failed unexpectedly due to max path limitation of NuGet package cache
I am using version `2.1.300-preview2-008251` of the SDK and I am getting this failure multiple times when building inside VS. Building from the command line completes successfully.

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	MSB4018	The "ResolvePackageAssets" task failed unexpectedly.
System.ArgumentNullException: Value cannot be null.
Parameter name: path1
   at System.IO.Path.Combine(String path1, String path2)
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.RaisePackageAssets[T](LockFileTarget target, Func`2 getAssets, Func`2 filter, Action`2 setup)
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.ExecuteCore()
   at Microsoft.NET.Build.Tasks.TaskBase.Execute()
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()	SocialWeather	C:\Users\James\.dotnet\x64\sdk\2.1.300-preview2-008251\Sdks\Microsoft.NET.Sdk\build\Microsoft.PackageDependencyResolution.targets	201	
```

Failing source: https://github.com/aspnet/SignalR/commit/4b666dfa856a7600205ab488a7ad4c8facbe7f6e

Open SignalR.sln in 15.6.2 and rebuild solution for the error.

[buildlogs.zip](https://github.com/dotnet/sdk/files/1812858/buildlogs.zip)

### PR DESCRIPTION
**Pull request by nguerrera**
Fix #2044

### CODE REVIEW

File: src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantDiagnosticsWhenPackageCannotBeFound.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                 .Pass();
+        }
+
+        private static void RemovePackageFromCache(TestPackageReference package)
+        {
+            var shaFile = Path.Combine(
+               TestContext.Current.NuGetCachePath,
+               package.ID,
+               package.Version,
+               $"{package.ID}.{package.Version}.nupkg.sha512");
```
Comment (dsplaisted) [inline src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantDiagnosticsWhenPackageCannotBeFound.cs]: Add comment here explaining why deleting the SHA works here.

**Review by dsplaisted — APPROVED** [review]

**Comment by nguerrera** [comment]
@MattGertz for approval to take to ship room for RTM

## Issues fixed

#2044

## Description of Issue

If packages go missing from the nuget cache for any reason between the time that the assets file is written by restore and when we build the assets cache, then there will be a crash in the build task. This can happen if the user clears his cache between restore and first build. It also effectively happens if restore hits a max path limitation. 

In fixing and testing this, also discovered another general issue where any crash during the assets cache generation will break subsequent incremental builds.

## Customer Impact

Unhelpful, scary crash instead of a useful error message that indicates the problem. 

Cannot recover from prior build crash without cleaning out intermediate directory.


## Risk

Low - Turns a definite crash case into an error message and delays writing the cache format signature until the cache is finalized.

## Testing

Regression coverage added as a unit test

This PR implements an integration test that would have caught this failure.  Manual tests were also run on macOS, Linux, and Windows to validate the change.

**Comment by MattGertz** [comment]
@nguerrera The crash and its ramifications are pretty bad, and on that basis alone I'd want to take it to Shiproom (and have changed the labels appropriately), but do we have a sense that this is a common thing, or just the one report in #2044?  We should understand that for Shiproom.

File: src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantDiagnosticsWhenPackageCannotBeFound.cs (hunk 1 above)
Comment (nguerrera) [reply src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantDiagnosticsWhenPackageCannotBeFound.cs]: Done in 315e27a

**Pull request merged**